### PR TITLE
any-wrap for opaque structs

### PR DESCRIPTION
**Pull request by bennn**
any-wrap now prohibits unsafe values & uses `chaperone-struct` in the fall-through case.

Questions:
- Are all the `base-val?` in `any-wrap.rkt` really safe?
- Are all the `unsafe-val?` in `any-wrap.rkt` really unsafe?
- I haven't been successful making example values for each base type. See lines 28 and 231 of the `variation-5` test. Any suggestions? Especially for the missing numeric types.

### CODE REVIEW

**Comment by bennn** [comment]
(I figured making this pull request would be the easiest way to update [#248](https://github.com/racket/typed-racket/pull/248). Let me know if I should move this.)

**Comment by samth** [comment]
Awesome, thanks! Looking now.

File: typed-racket-lib/typed-racket/utils/any-wrap.rkt [hunk 1]
```diff
@@ -1,21 +1,50 @@
...
 (define (base-val? e)
   (or (number? e) (string? e) (char? e) (symbol? e)
-      (null? e) (regexp? e) (eq? undefined e) (path? e)
-      (regexp? e) (keyword? e) (bytes? e) (boolean? e) (void? e)
+      (null? e) (eq? undefined e) (path? e) (eof-object? e)
+      (regexp? e) (pregexp? e) (byte-regexp? e) (byte-pregexp? e)
+      (keyword? e) (bytes? e) (boolean? e) (void? e)
+      (continuation-prompt-tag? e) (continuation-mark-key? e)
+      (module-path? e) (resolved-module-path? e)
+      (pretty-print-style-table? e) (namespace-anchor? e)
```
Comment (samth) [inline typed-racket-lib/typed-racket/utils/any-wrap.rkt]: I think namespace-anchors are an issue (see the typed racket issue I just filed).

File: typed-racket-lib/typed-racket/utils/any-wrap.rkt [hunk 2]
```diff
@@ -1,21 +1,50 @@
...
+      (variable-reference? e) (impersonator-property? e)
```
Comment (samth) [inline typed-racket-lib/typed-racket/utils/any-wrap.rkt]: variable references have the same problem as namespace-anchors.

File: typed-racket-lib/typed-racket/utils/any-wrap.rkt [hunk 3]
```diff
@@ -1,21 +1,50 @@
...
+      (semaphore? e) (fsemaphore? e)
+      (bytes-converter? e)
+      (pseudo-random-generator? e)
+      (logger? e)
+      (continuation-mark-set? e)
```
Comment (samth) [inline typed-racket-lib/typed-racket/utils/any-wrap.rkt]: A continuation mark set can contain a function.

File: typed-racket-lib/typed-racket/utils/any-wrap.rkt [hunk 4]
```diff
@@ -1,21 +1,50 @@
...
       ;; Base values because you can only store flonums/fixnums in these
       ;; and not any higher-order values. This isn't sound if we ever
       ;; introduce bounded polymorphism for Flvector/Fxvector.
-      (flvector? e) (fxvector? e) (extflvector? e)))
+      (flvector? e) (fxvector? e) (extflvector? e) (extflonum? e)))
+
+(define (unsafe-val? e)
+  (or (syntax? e) (namespace? e) (weak-box? e) (ephemeron? e)
+      (mpair? e) (thread-cell? e) (class? e) (unit? e)
+      (compiled-module-expression? e) (compiled-expression? e)
```
Comment (samth) [inline typed-racket-lib/typed-racket/utils/any-wrap.rkt]: How can you smuggle a higher-order value via a compiled expression?

File: typed-racket-lib/typed-racket/utils/any-wrap.rkt [hunk 5]
```diff
@@ -1,21 +1,50 @@
...
+      (struct-type-property? e)
+      (special-comment? e)
+      (udp? e) (custodian? e) (parameterization? e)
```
Comment (samth) [inline typed-racket-lib/typed-racket/utils/any-wrap.rkt]: I think `udp?` should be safe.

File: typed-racket-lib/typed-racket/utils/any-wrap.rkt [hunk 6]
```diff
@@ -1,21 +1,50 @@
...
+      (internal-definition-context? e)
+      (thread-group? e)
+      (inspector? e)
```
Comment (samth) [inline typed-racket-lib/typed-racket/utils/any-wrap.rkt]: I think inspectors are safe.

File: typed-racket-lib/typed-racket/utils/any-wrap.rkt [hunk 7]
```diff
@@ -1,21 +1,50 @@
...
+
+(define (unsafe-val? e)
+  (or (syntax? e) (namespace? e) (weak-box? e) (ephemeron? e)
+      (mpair? e) (thread-cell? e) (class? e) (unit? e)
+      (compiled-module-expression? e) (compiled-expression? e)
+      (struct-type-property? e)
+      (special-comment? e)
+      (udp? e) (custodian? e) (parameterization? e)
+      (internal-definition-context? e)
+      (thread-group? e)
```
Comment (samth) [inline typed-racket-lib/typed-racket/utils/any-wrap.rkt]: What can you do with a thread-group?

File: typed-racket-test/succeed/pr241-variation-5.rkt [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+;; These USE functions should exercise specific functionality
+;;  and not be dummy functions like (lambda (x) x).
+(define-for-syntax known-base-types '(
+ ;; --- TODO missing value for these types
+ ;[ExtFlonum-Zero (->extfl 0) extflround]
+ ;[ExtFlonum-Negative-Zero (->extfl -0) extflround]
+ ;[ExtFlonum-Positive-Zero (->extfl 0) extflround]
+ ;[Single-Flonum-Complex 1f0+1i add1]
+ ;[Undefined undefined boolean?]
+ ;[Nothing (Un)]
```
Comment (samth) [inline typed-racket-test/succeed/pr241-variation-5.rkt]: The ones above this are all very simple base types, and we don't need to worry.

File: typed-racket-test/succeed/pr241-variation-5.rkt [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+ ;[Prompt-TagTop (default-continuation-prompt-tag) continuation-prompt-available?]
+ ;[Struct-TypeTop (let () (struct s ()) (s)) (lambda (x) x)]
+ ;[Module-Path-Index (let ([x 4]) (variable-reference->module-path-index x)) module-path-index-resolve]
```
Comment (samth) [inline typed-racket-test/succeed/pr241-variation-5.rkt]: Also a base type.

File: typed-racket-test/succeed/pr241-variation-5.rkt [hunk 10]
```diff
@@ -0,0 +1,339 @@
...
+;;   [TYPE-TAG TYPE VALUE USE]
+;; TYPE-TAG is a polymorphic type exported by Typed Racket
+;; TYPE is an instantiation of TYPE-TAG
+;; VALUE is an expression with type TYPE
+;; USE is the same as for `known-base-types`
+(define-for-syntax known-poly-types '(
+  ;; --- TODO need value for these types
+  ;[Async-Channelof (Async-Channelof Any) (make-async-channel) async-channel-try-get]
+  ;[Prompt-Tagof (Prompt-Tagof Any Any) (default-continuation-prompt-tag) #f]
+  ;[Evtof (Evtof String) (make-log-receiver (current-logger)) choice-evt]
```
Comment (samth) [inline typed-racket-test/succeed/pr241-variation-5.rkt]: `evt`s are chaperonable.

File: typed-racket-test/succeed/pr241-variation-5.rkt [hunk 11]
```diff
@@ -0,0 +1,339 @@
...
+;; `known-polymorphic-types` is a list of 4-lists:
+;;   [TYPE-TAG TYPE VALUE USE]
+;; TYPE-TAG is a polymorphic type exported by Typed Racket
+;; TYPE is an instantiation of TYPE-TAG
+;; VALUE is an expression with type TYPE
+;; USE is the same as for `known-base-types`
+(define-for-syntax known-poly-types '(
+  ;; --- TODO need value for these types
+  ;[Async-Channelof (Async-Channelof Any) (make-async-channel) async-channel-try-get]
+  ;[Prompt-Tagof (Prompt-Tagof Any Any) (default-continuation-prompt-tag) #f]
```
Comment (samth) [inline typed-racket-test/succeed/pr241-variation-5.rkt]: Prompt tags have chaperones.

File: typed-racket-test/succeed/pr241-variation-5.rkt [hunk 12]
```diff
@@ -0,0 +1,339 @@
...
+
+;; `known-polymorphic-types` is a list of 4-lists:
+;;   [TYPE-TAG TYPE VALUE USE]
+;; TYPE-TAG is a polymorphic type exported by Typed Racket
+;; TYPE is an instantiation of TYPE-TAG
+;; VALUE is an expression with type TYPE
+;; USE is the same as for `known-base-types`
+(define-for-syntax known-poly-types '(
+  ;; --- TODO need value for these types
+  ;[Async-Channelof (Async-Channelof Any) (make-async-channel) async-channel-try-get]
```
Comment (samth) [inline typed-racket-test/succeed/pr241-variation-5.rkt]: We could implement async channel contracts.

File: typed-racket-lib/typed-racket/utils/any-wrap.rkt (hunk 4 above)
Comment (bennn) [inline typed-racket-lib/typed-racket/utils/any-wrap.rkt]: I thought I could do it by decompiling the expression, but haven't been able to yet. Marking as "safe" for now.

File: typed-racket-test/succeed/pr241-variation-5.rkt (hunk 11 above)
Comment (bennn) [inline typed-racket-test/succeed/pr241-variation-5.rkt]: Do they have chaperones by default, or do we need to make a chaperone in `any-wrap`?

**Comment by bennn** [comment]
What do you think of this `compile` abuse? Should we prohibit compiled expressions as `Any`?

```
#lang racket/base

(module u racket/base
  (define (foo? x)
    (define-values (ctr incr) (eval x))
    (set-box! ctr 'haha)
    (displayln (incr))
    #t)
  (provide foo?))

(module t typed/racket/base
  (require/typed (submod ".." u)
    [#:opaque T foo?])
  (define compiled-counter
    (compile-syntax
      #'(let*: ([b : (Boxof Integer) (box 0)]
                [incr : (-> Void) (lambda () (set-box! b (add1 (unbox b))))])
          (values b incr))))
  (foo? b))

(require 't)
```

**Comment by samth** [comment]
Is the trick with compile any worse than what we can already do with syntax objects?

**Comment by bennn** [comment]
I don't think so, but syntax objects also aren't allowed as `Any`.

**Comment by samth** [comment]
Sorry for leaving this for a while. What I mean is that you can provide a syntax object as `Syntax`, which allows for the same problems.

File: typed-racket-lib/typed-racket/utils/any-wrap.rkt [hunk 13]
```diff
@@ -1,21 +1,63 @@
...
   (or (number? e) (string? e) (char? e) (symbol? e)
-      (null? e) (regexp? e) (eq? undefined e) (path? e)
-      (regexp? e) (keyword? e) (bytes? e) (boolean? e) (void? e)
+      (null? e) (eq? undefined e) (path? e) (eof-object? e)
+      (regexp? e) (pregexp? e) (byte-regexp? e) (byte-pregexp? e)
+      (keyword? e) (bytes? e) (boolean? e) (void? e)
+      (bytes-converter? e)
+      (compiled-expression? e)
+      (compiled-module-expression? e)
+      (continuation-prompt-tag? e) ;;bg; apply a chaperone?
```
Comment (samth) [inline typed-racket-lib/typed-racket/utils/any-wrap.rkt]: I don't think this is safe by itself -- it allows you to call a prompt handler.

File: typed-racket-lib/typed-racket/utils/any-wrap.rkt [hunk 14]
```diff
@@ -1,21 +1,63 @@
...
+      (custodian? e)
+      (ephemeron? e)
+      (future? e)
+      (internal-definition-context? e)
+      (mpair? e)
+      (namespace-anchor? e)
+      (namespace? e)
+      (parameterization? e)
+      (security-guard? e)
+      (special-comment? e)
```
Comment (samth) [inline typed-racket-lib/typed-racket/utils/any-wrap.rkt]: I think we can just wrap `special-comment` values ourselves.

File: typed-racket-test/succeed/pr241-variation-5.rkt [hunk 15]
```diff
@@ -0,0 +1,344 @@
...
+))
+
+;; We don't have values to test these types with, but trust they're wrapped correctly
+(define-for-syntax whitelist '(
+ (Undefined . #f)
+ (Nothing . #f)
+ (Struct-TypeTop . #f)
+ (Module-Path-Index . #f)
+ (Compiled-Module-Expression . #f)
+ (Namespace-Anchor . "See `variation-6` for this pull request")
```
Comment (samth) [inline typed-racket-test/succeed/pr241-variation-5.rkt]: Include an actual URL to the PR here.

**Comment by bennn** [comment]
(Still needs to handle special comments & async channels)

**Comment by samth** [comment]
Let's just treat those two as unsafe and move forward with this.

**Pull request merged**